### PR TITLE
Rec experimental qname minimization

### PR DESCRIPTION
**Pull request by omoerbeek**
### Short description

Only tested in a local environment.
It implements a relaxed form of qname minimization, based on A queries with fallback to the full name if something goes wrong.

Enable with config option qname-minimization=yes (default no)

### Checklist

I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] compiled this code
- [X] tested this code
- [ ] included documentation (including possible behaviour changes)
- [X] documented the code
- [ ] added or modified regression test(s)
- [X] added or modified unit test(s)

### CODE REVIEW

**Comment by omoerbeek** [comment]
IMO this is now good for wider consumption. The current state of the algorithm and code is simpler than the original commit.

File: pdns/recursordist/docs/settings.rst [hunk 1]
```diff
@@ -1164,6 +1164,17 @@ Whether to compute the latency of responses in protobuf messages using the times
...
 
+.. _setting-qname-minimisation:
+
+``qname-minimisation``
+----------------------
+.. versionadded:: 4.3.0
+-  Boolean
+-  Default: no
+
+Enable Query Name Minimisation. This is a experimental feature, implememting a relaxed form of Query Name Mimimisation as
```
Comment (zeha) [inline pdns/recursordist/docs/settings.rst]: `implementing`

File: pdns/syncres.cc [hunk 2]
```diff
@@ -540,6 +541,125 @@ int SyncRes::asyncresolveWrapper(const ComboAddress& ip, bool ednsMANDATORY, con
...
+    bool flawedNSSet = false;
+    set<GetBestNSAnswer> beenthereIgnored;
+    getBestNSFromCache(qname, qtype, bestns, &flawedNSSet, depth + 1, beenthereIgnored);
+
+    DNSName ancestor;
+    if (bestns.size() > 0) {
+      ancestor = bestns[0].d_name;
+      QLOG("Step1 Ancestor from cache is " << ancestor.toString());
+    } else {
+      QLOG("Step1 No ancestor found, fall back to regular resolve");
```
[the remark this answers is not in the document]
Comment (omoerbeek) [reply pdns/syncres.cc]: Fixed in upcoming commiit

File: pdns/recursordist/docs/settings.rst [hunk 3]
```diff
@@ -1164,6 +1164,17 @@ Whether to compute the latency of responses in protobuf messages using the times
...
+.. _setting-qname-minimization:
+
+``qname-minimization``
+----------------------
+.. versionadded:: 4.3.0
+-  Boolean
+-  Default: no
+
+Enable Query Name Minimization. This is a experimental feature, implementing a relaxed form of Query Name Mimimization as
+described in RFC 7816.
```
Comment (pieterlexis) [inline pdns/recursordist/docs/settings.rst]: This will make a nice link to the RFC:

```suggestion
described in :rfc:`7816`.
```

**Pull request merged**